From: Martin Schwidefsky <schwidefsky@de.ibm.com>
Date: Fri, 20 Apr 2018 12:48:52 +0200
Subject: s390: correct module section names for expoline code revert
References: bsc#1089386
Git-commit: 6cf09958f32b9667bb3ebadf74367c791112771b
Patch-mainline: v4.17-rc3

The main linker script vmlinux.lds.S for the kernel image merges
the expoline code patch tables into two section ".nospec_call_table"
and ".nospec_return_table". This is *not* done for the modules,
there the sections retain their original names as generated by gcc:
".s390_indirect_call", ".s390_return_mem" and ".s390_return_reg".

The module_finalize code has to check for the compiler generated
section names, otherwise no code patching is done. This slows down
the module code in case of "spectre_v2=off".

Cc: stable@vger.kernel.org # 4.16
Fixes: f19fbd5ed6 ("s390: introduce execute-trampolines for branches")
Signed-off-by: Martin Schwidefsky <schwidefsky@de.ibm.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 arch/s390/kernel/module.c |    4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

--- a/arch/s390/kernel/module.c
+++ b/arch/s390/kernel/module.c
@@ -459,10 +459,10 @@ int module_finalize(const Elf_Ehdr *hdr,
 			apply_alternatives(aseg, aseg + s->sh_size);
 
 #ifdef CONFIG_EXPOLINE
-		if (!strcmp(".nospec_call_table", secname))
+		if (!strncmp(".s390_indirect", secname, 14))
 			nospec_revert(aseg, aseg + s->sh_size);
 
-		if (!strcmp(".nospec_return_table", secname))
+		if (!strncmp(".s390_return", secname, 12))
 			nospec_revert(aseg, aseg + s->sh_size);
 #endif
 	}
